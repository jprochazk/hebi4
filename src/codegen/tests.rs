use std::{fs::read_to_string, path::Path};

use super::emit;
use crate::{codegen::EmitOptions, parser::parse, token::tokenize};

fn parse_options(options: &str) -> EmitOptions {
    let mut o = EmitOptions::empty();

    for opt in options.split(',') {
        match opt {
            "dce" => o.dead_code_elimination = true,
            _ => {}
        }
    }

    o
}

#[glob_test::glob("../../tests/inputs/codegen/*.hi")]
fn emitter(path: &Path) {
    let input = read_to_string(path).unwrap();
    let input = input.trim();
    let options = if input.starts_with("//") {
        let first_line = input.lines().next().unwrap().trim_start_matches("//");
        parse_options(first_line)
    } else {
        Default::default()
    };

    let tokens = tokenize(input);
    let ast = match parse(&tokens) {
        Ok(ast) => ast,
        Err(err) => {
            panic!("{}", err.render(input));
        }
    };
    let (snapshot, failure) = match emit("test".into(), &ast, options) {
        Ok(m) => (format!("SOURCE\n{input}\n\nOK\n{}", m.disasm(input)), false),
        Err(err) => (
            format!("SOURCE\n{input}\n\nERROR\n{}", err.render(input)),
            true,
        ),
    };

    #[cfg(not(miri))]
    {
        insta::assert_snapshot!(snapshot);
        let _ = failure;
    }

    #[cfg(miri)]
    {
        if failure {
            panic!("{snapshot}");
        } else {
            eprintln!("{snapshot}");
        }
    }
}

// NOTE: leave this here, it's handy for debugging:
// 1. uncomment this test
// 2. filter for it (`cargo test -- _temp`),
// 3. set a breakpoint somewhere in codegen

// #[test]
// fn _temp() {
//     emitter(std::path::Path::new("tests/inputs/run/arithmetic_add.hi"));
//     panic!();
// }
