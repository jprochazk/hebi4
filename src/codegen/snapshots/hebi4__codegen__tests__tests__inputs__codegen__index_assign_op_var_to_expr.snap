---
source: src/codegen/tests.rs
expression: snapshot
---
SOURCE
let o = {a:0}
fn k() { "a" }
let v = 1
o[k()] = o[k()] + v
o[k()] += v

OK
.fn @main
  params 0
  stack 6
  .locals
    o = r1
    v = r2
  .code
   0  ltable r1, 1
   1  lsmi r2, 0
   2  skeyc r1, "a", r2
   3  lsmi r2, 1
   4  call r3, fn1   ; fn1=k
   5  call r5, fn1   ; fn1=k
   6  lidx r4, r1, r5
   7  add r4, r4, r2
   8  sidx r1, r3, r4
   9  call r3, fn1   ; fn1=k
  10  lidx r4, r1, r3
  11  add r4, r4, r2
  12  sidx r1, r3, r4
  13  lnil r0
  14  ret

.fn k
  params 0
  stack 1
  .code
  0  lstr r0, "a"
  1  ret
